sys/fmt/fmt.c

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ * @ingroup sys_fmt
+ * @{
+ * @file
+ * @brief String formatting library implementation
+ * @author Kaspar Schleiser <kaspar@schleiser.de>
+ * @}
+ */
+
+#include <assert.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <unistd.h>
+
+#include "div.h"
+#include "fmt.h"
+
+static const char _hex_chars[16] = "0123456789ABCDEF";
+
+static inline int _is_digit(char c)
+{
+    return (c >= '0' && c <= '9');
+}
+
+size_t fmt_byte_hex(char *out, uint8_t byte)
+{
+    if (out) {
+        *out++ = _hex_chars[byte >> 4];
+        *out = _hex_chars[byte & ~0xF0];
+    }
+    return 2;
```

Comment:
It would... In this case, I went for consistency over saving that instruction. ("All fmt_* functions return the number of bytes that would have been written if out parameter is NULL.")

Also, AFAIK C promotes parameters < void* to wordsize when passing parameters to functions, so a printf implementation using fmt_* as backend could more consistently use fmt_* as backend functions if they have equal signatures.

And, IMHO lto will sort that out eventually anyways. ;)